File: 132-pattern.py
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        N = len(nums)
        mins = [0]*N
        min_ = nums[0]
        for i,n in enumerate(nums):
            if n < min_:
                min_ = n
            mins[i] = min_
        stk = []
        for i in range(N - 1, 0, -1):
            if stk and mins[i] < stk[-1] < nums[i]:
                return True
            stk.append(nums[i])
            while stk and stk[-1] <= mins[i-1]:
                stk.pop()
        return False

File: after-camp/132-pattern.cpp
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        int N = nums.size();
        vector<int> mins(N);
        mins[0] = nums[0];
        for(int i = 1; i < N; ++i) mins[i] = min(mins[i - 1], nums[i]);
        stack<int> monoStk; monoStk.push(INT_MAX);
        for(int i = N - 1; i >= 0; --i){
            while(mins[i] >= monoStk.top()) monoStk.pop();
            if(nums[i] > monoStk.top() && monoStk.top() > mins[i]) return true;
            if(monoStk.top() > nums[i]) monoStk.push(nums[i]);
        }
        return false;
    }
};